### PR TITLE
Fix artifact completion in <dependencies> in VSCode

### LINKED ISSUE #180: Completion for partially complete artifact names in dependencies
Currently, you can add a dependency by opening completion at the `|`, typing the artifact name, then selecting which one you want:
```xml
...
  <dependencies>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-core</artifactId>
      <version>2.11.3</version>
    </dependency>
    |
  </depdenencies>
...
```

It would be helpful if completion could also be provided if you type part of the artifact name, then open completion (at the `|`):
```xml
...
  <dependencies>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-core</artifactId>
      <version>2.11.3</version>
    </dependency>
    junit-|
  </depdenencies>
...
```

It would also be helpful if completion could be provided inside of `<dependencies>` before any other dependencies are declared:
```xml
...
  <dependencies>
    junit-|
  </depdenencies>
...
```

### PR DESCRIPTION
**Pull request by datho7561**
Adjust the replace range of the TextEdit of the completion item so that it covers just the text node. This prevents VSCode from filtering the completion result.

Closes #180

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
- [ ] Tests
- [ ] Take a look for regressions, for instance in Eclipse

**Comment by angelozerr** [comment]
Thanks @datho7561 for your PR. For the test, it should be nice to consume LemMinx XMLAssert.

**Comment by mickaelistria** [comment]
@datho7561 Thanks for the move to JUnit 5! Do you think you could make it a separate dedicated PR; we could merge it almost immediately.

**Comment by datho7561** [comment]
@mickaelistria #187

**Comment by mickaelistria** [comment]
@datho7561 Would it be possible to stick to LemMinX 0.14.1 for this change or are there API changes requiring the move to 0.14.2-SNAPSHOT? We may want to release LemMinX-Maven after a 1st batch of issue to have it working in VSCode is resolved; and we need to keep things compatible with Eclipse IDE, which at the moment is using 0.14.1 and may not be able to update to 0.14.2 as soon as it's released.

**Comment by datho7561** [comment]
> ...are there API changes requiring the move to 0.14.2-SNAPSHOT? ...

There aren't API changes that are required, sorry I copy/pasted the dependency info from somewhere else and haven't fixed it in this branch yet.

> We may want to release LemMinX-Maven after a 1st batch of issue to have it working in VSCode is resolved; and we need to keep things compatible with Eclipse IDE, which at the moment is using 0.14.1 and may not be able to update to 0.14.2 as soon as it's released.

Okay sounds good.

**Comment by mickaelistria** [comment]
I'll try it tomorrow morning in Eclipse IDE and will merge if I don't see regressions.

**Comment by datho7561** [comment]
Okay. The test that I added fails as of right now.

**Comment by mickaelistria** [comment]
So I'll wait for the test to pass before trying it in Eclipse IDE.
Do you have any clue why the test seems to not be run with https://github.com/eclipse/lemminx-maven/pull/182/checks?check_run_id=1683898573 while it's running with Jenkins?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenCompletionParticipant.java [hunk 1]
```diff
@@ -432,16 +434,34 @@ private CompletionItem toGAVCompletionItem(ArtifactInfo artifactInfo, ICompletio
 		if (artifactInfo.getDescription() != null) {
 			item.setDocumentation(artifactInfo.getDescription());
 		}
+		Range newReplaceRange;
+		try {
+			DOMDocument doc = request.getXMLDocument();
```
Comment (angelozerr) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenCompletionParticipant.java:439]: In LemMinx we had a similar usecase when you try to open completion for element tag name inside a text node which doesn't start with `<`. See https://github.com/eclipse/lemminx/blob/0d812fb4acca54b9df17371e0bb8e8b99bcf5b73/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLCompletions.java#L755

If you see `getElementNameRangeAt `https://github.com/eclipse/lemminx/blob/0d812fb4acca54b9df17371e0bb8e8b99bcf5b73/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocument.java#L125 it consumes getWordRangeAt which use a regexp https://github.com/eclipse/lemminx/blob/0d812fb4acca54b9df17371e0bb8e8b99bcf5b73/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/TextDocument.java#L95

I suggest that you write a regexp pattern for artifact id and you do 

`Range artifactIdRange = doc.getTextDocument(request.getOffset(), ARTIFACTID_PATTERN);`

(artifactdIdRange is your newReplaceRange )

Comment (datho7561) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenCompletionParticipant.java]: Thanks for pointing this out! I have switched to this method for the PR

**Comment by datho7561** [comment]
> So I'll wait for the test to pass before trying it in Eclipse IDE.

They should be good now.

> Do you have any clue why the test seems to not be run with https://github.com/eclipse/lemminx-maven/pull/182/checks?check_run_id=1683898573 while it's running with Jenkins?

I'll take a look into it.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/test/LemminxMavenCompletionParticipantTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+import static org.eclipse.lemminx.XMLAssert.c;
+import static org.eclipse.lemminx.XMLAssert.te;
+
+import org.eclipse.lemminx.XMLAssert;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(NoMavenCentralIndexExtension.class)
+public class LemminxMavenCompletionParticipantTest {
```
Comment (angelozerr) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/test/LemminxMavenCompletionParticipantTest.java:19]: I would rename the class with just `MavenCompletionTest`

**Comment by angelozerr** [comment]
Test on Windows OS with vscode and it works like a charm, great job @datho7561 !

**Comment by mickaelistria** [comment]
Tested on m2e and seems to work as expected, no regression found.
Thanks @datho7561 !

**Pull request merged**